src/github.com/getlantern/idletiming/idletiming_conn.go

```diff
@@ -100,12 +105,13 @@ func (c *IdleTimingConn) Read(b []byte) (int, error) {
 			n, err := c.conn.Read(b)
 			c.markActive(n)
 			totalN = totalN + n
-			timedOut := isTimeout(err)
-			if timedOut {
-				// Ignore timeouts when using deadline based on IdleTimeout
+			hitMaxDeadline := isTimeout(err) && !time.Now().Before(maxDeadline)
```

Comment:
!time.Now().Before(maxDeadline) is important.  This controls for the situation where the read/write times out not because of the deadline that we set above but for some other reason.  In that case, we want to just go ahead and return the timeout error and break out of the loop.